runtime/j9vm/jvm.c

```diff
@@ -4544,34 +4633,42 @@ JVM_RegisterSignal(jint sigNum, void* handler)
 
 	Trc_SC_RegisterSignal();
 
-	if (isSignalUsedByVM(sigNum)) {
-		/* Don't allow user to register a native handler since
-		 * the signal is already used by the VM.
-		 */
-	} else if (J9_ARE_NO_BITS_SET(javaVM->sigFlags, J9_SIG_XRS_SYNC)
-			&& isSignalSpecial(sigNum)) {
-		/* Don't allow user to register a native handler since
-		 * the signal is already used by the VM.
+	if (isSignalReservedByJVM(sigNum)) {
```

Comment:
`JVM_RegisterSignal` registers a Java `SignalHandler` with a signal. When the reserved signals are unreserved by specifying the `-Xrs` option, the spec says that the user can register their own `native handler`. `JVM_RegisterSignal` doesn't allow a user to register a `_native_handler`; it registers a Java `SignalHandler`. So, we shouldn't let a user register Java `SignalHandlers` via `JVM_RegisterSignal` with the reserved signals ... when the reserved signals are unreserved by specifying `-Xrs` option. The users should use a native approach to register a `native handler` for the reserved signals. @pshipton Do you agree with my interpretation?

Reply:
Makes sense. How about adding this information to the JVM_RegisterSignal doc.

The whole implementation seems overly complicated for something that seems to only need to handle 3 signals: HUP, INT, TERM.

Reply:
We need to support more than four signals. Referring to [SunMiscSignalTest](https://github.ibm.com/runtimes/openj9-openjdk-jdk9/blob/ibm_sdk/jdk/test/sun/misc/SunMiscSignalTest.java#L101). You should see a 2d array. The fist column is signal name, the second column determines if that signal is supported, the third column determines if we can register a handler via `JVM_RegisterSignal`, the fourth column determines if that signal can be raised via `JVM_RaiseSignal` and the fifth column determines if the associated handler should be invoked.

Reply:
This implementation doesn't seem to match the test suite, which I believe shows that (for example) SIGBUS can be registered, raised, and the java handler invoked (unless running with -Xrs).

Reply:
It won't. List of reserved signals will vary between Hotspot and OpenJ9. In OpenJ9, `SIGBUS` is reserved by the JVM. It is handled as a synchronous signal via `omrsig_protect`.

Reply:
Since OpenJ9 is supporting synchronous signals differently then we should consider simplifying this implementation to the minimum required to support the asynchronous signals.

Reply:
Updated `JVM_RegisterSignal` and `JVM_RaiseSignal` doc to match the current implementation.

Currently, `JVM_*Signal` functions only support asynchronous signals. Synchronous signals are ignored or an error is returned for them (`...: Signal already used by VM or OS`). OpenJ9's error handling won't match Hotspot's error handling if we exclude synchronous signals. There won't be much simplification; only the list of reserved signals will reduce in `isSignalReservedByJVM()` and we may get rid of the five synchronous signals in `jvm.c::signalMap`.